cmd/tidb-lightning/main.go

```diff
@@ -96,3 +105,47 @@ func main() {
 		os.Exit(1)
 	}
 }
+
+func checkSystemRequirement(cfg *config.GlobalConfig) error {
+	if cfg.TikvImporter.Backend == config.BackendLocal {
+		// in local mode, we need to read&write a lot of L0 sst files, so we need to check
+		// system max open files limit
+		var totalSize uint64
+		err := filepath.Walk(cfg.Mydumper.SourceDir, func(path string, info os.FileInfo, err error) error {
+			if err != nil {
+				return err
+			}
+			if !info.IsDir() && info.Size() > 0 {
+				totalSize += uint64(info.Size())
+			}
+			return err
+		})
+		if err != nil {
+			return errors.Trace(err)
+		}
+
+		// we estimate each sst as 50MB, this is a fairly coarse predict, the actual fd needed is
+		// depend on chunk&import concurrency and each data&index key-value size
+		estimateMaxFiles := totalSize / (50 * 1024 * 1024)
```

Comment:
how this "50 MiB" is determined? 

also, for a 10 TB data set this means Pebble will open up to 200000 files simultaneously. This sounds more like a bug to me (that the engines are not closed quickly enough).

Reply:
In the tpcc benchmaerk, with a memtable size of 512MB, the generated L0 sst file sizes ranges from 40~450MB, so I'm not sure how to estimate the avg size. And In the common case, L0 sst files will be compressed to L1~Ln level ssts, thus total files should be very large, but we prevent the compression phase in favor of better performance thus the generated files numbers will be quite big. Because index engine can't be split, so if a table is huge, and there are many indices, then maybe we will generate thousands of L0 sst files.

Reply:
can we use the largest N total size of each table (N = index concurrency) instead of the entire data source?

from the pebble source code it seems 50 MiB = [25](https://github.com/cockroachdb/pebble/blob/a452ae2c916a74a035d264e42c9b761061a988b5/compaction.go#L41) × [2 MiB](https://github.com/cockroachdb/pebble/blob/a452ae2c916a74a035d264e42c9b761061a988b5/options.go#L225), so maybe we could increase the TargetFileSize? 🤔 In TiKV Importer [this is set](https://github.com/tikv/importer/blob/9531815ad95fac8dc6febcf0fb8cd76d01662df3/src/import/engine.rs#L385) to [1 GiB](https://github.com/tikv/importer/blob/9531815ad95fac8dc6febcf0fb8cd76d01662df3/etc/tikv-importer.toml#L23).

Reply:
Seems the L0 sst file size is depend on the memory table size and the conf `MaxOpenFiles` is used for compaction. After explict set `MaxOpenFiles` option to 1GB, the table `order_line` index engine still generates a lots of sst files with size of 54MB. And almost each engine generates sst files with different size.  Maybe we should use `MemoryTableSize` as each sst file's size, as memory table accounts for the raw size of the kv-pairs.
So we can estimate by:  {top N table size} / {MemoryTableSize} as the needed fds

Reply:
Okay.